File: 2.OOP/concepts/MediumExample/design_patterns/Creational/Lazy-initialization/Lazy-initialization.go
package main

import (
	"fmt"
	"time"
)

type Database struct {
	connection chan *Database
}

func newDatabase() *Database {
	return &Database{connection: make(chan *Database)}
}

func (db *Database) getConnectionMessage() string {
	return "Database connection established"
}

func (db *Database) getConnection() string {
	return db.getConnectionMessage()
}

func (db *Database) createConnection() {
	time.Sleep(1 * time.Second)
	db.connection <- db
}

func main() {
	db := newDatabase()
	go func() {
		for {
			db.createConnection()
		}
	}()
	for i := 0; i < 2; i++ {
		fmt.Println(db.getConnection())
	}
}

File: 2.OOP/concepts/MediumExample/design_patterns/Creational/Lazy-initialization/Lazy-initialization.ts
class Database {
  private static connection: Promise<Database> | null = null;

  private constructor() {}

  // Database connection simulation
  static async getConnection(): Promise<string> {
    if (!this.connection) {
      console.log("Connecting...");
      try {
        this.connection = new Promise((resolve, reject) => {
          setTimeout(() => {
            resolve(new Database());
          }, 1000);
        });
      } catch (error) {
        console.error("Error establishing database connection:", error);
        throw error; 
      }
    }

    return (await this.connection).getConnectionMessage();
  }

  private getConnectionMessage(): string {
    return "Database connection established";
  }
}

(async () => {
  try {
    const connection1 = await Database.getConnection();
    console.log(connection1);

    const connection2 = await Database.getConnection();
    console.log(connection2);
  } catch (error) {
    console.error("An error occurred while using the database:", error);
  }
})();
